src/Compilers/CSharp/Portable/FlowAnalysis/VariablesDeclaredWalker.cs

```diff
@@ -122,16 +122,22 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
             return base.VisitLocalFunctionStatement(node);
         }
 
-        public override void VisitForEachIterationVariable(BoundForEachStatement node)
+        public override void VisitForEachIterationVariables(BoundForEachStatement node)
         {
             if (IsInside)
             {
-                if (node.IterationVariableOpt != null)
+                var deconstructionAssignment = node.DeconstructionOpt?.DeconstructionAssignment;
+
+                if (deconstructionAssignment == null)
                 {
-                    _variablesDeclared.Add(node.IterationVariableOpt);
+                    // regular, not deconstructing, foreach declares exactly one iteration variable
+                    _variablesDeclared.Add(node.IterationVariables.Single());
+                }
+                else
+                {
+                    // deconstruction foreach declares multiple variables
+                   deconstructionAssignment.Left.VisitAllElements((x, self) => self.Visit(x), this);
```

Comment:
This fix does not change the existing behavior.
Rationalizing how this scenario should work is likely to not meet RC3 bar so I have entered a separate issue: https://github.com/dotnet/roslyn/issues/16370 so that it could be triaged accordingly.